### ISSUE TITLE
1.8.1: build fails with cmake -Dgtest_build_tests=ON

### ISSUE THREAD

**Issue by kloczek** [google/googletest#2298]
<pre>[ 36%] <font color="#4E9A06"><b>Linking CXX executable gtest_unittest</b></font>
cd /home/tkloczko/rpmbuild/BUILD/googletest-release-1.8.1/x86_64-redhat-linux-gnu/googlemock/gtest &amp;&amp; /usr/bin/cmake -E cmake_link_script CMakeFiles/gtest_unittest.dir/link.txt --verbose=1
/usr/bin/g++  -O2 -g -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -fstack-protector-strong -grecord-gcc-switches -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1 -m64 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection -flto  -Wl,-z,relro   -Wl,-z,now -specs=/usr/lib/rpm/redhat/redhat-hardened-ld -flto -fuse-linker-plugin -rdynamic CMakeFiles/gtest_unittest.dir/test/gtest_unittest.cc.o  -o gtest_unittest -Wl,-rpath,/home/tkloczko/rpmbuild/BUILD/googletest-release-1.8.1/x86_64-redhat-linux-gnu/googlemock/gtest libgtest_main.so.1.8.1 libgtest.so.1.8.1 -pthread 
/home/tkloczko/rpmbuild/BUILD/googletest-release-1.8.1/googletest/test/gtest_unittest.cc: In function &apos;__static_initialization_and_destruction_0.constprop&apos;:
/home/tkloczko/rpmbuild/BUILD/googletest-release-1.8.1/googletest/test/gtest_unittest.cc:7795:1: note: variable tracking size limit exceeded with &apos;-fvar-tracking-assignments&apos;, retrying without
 7795 | }
      | ^
/usr/bin/ld: /tmp/gtest_unittest.GfacXD.ltrans19.ltrans.o: in function `testing::internal::StreamingListener::OnTestPartResult(testing::TestPartResult const&amp;)&apos;:
/home/tkloczko/rpmbuild/BUILD/googletest-release-1.8.1/googletest/src/gtest-internal-inl.h:1162: undefined reference to `testing::internal::StreamingListener::UrlEncode[abi:cxx11](char const*)&apos;
/usr/bin/ld: /home/tkloczko/rpmbuild/BUILD/googletest-release-1.8.1/googletest/src/gtest-internal-inl.h:1160: undefined reference to `testing::internal::StreamingListener::UrlEncode[abi:cxx11](char const*)&apos;
collect2: error: ld returned 1 exit status
make[2]: *** [googlemock/gtest/CMakeFiles/gtest_unittest.dir/build.make:89: googlemock/gtest/gtest_unittest] Error 1
make[2]: Leaving directory &apos;/home/tkloczko/rpmbuild/BUILD/googletest-release-1.8.1/x86_64-redhat-linux-gnu&apos;
make[1]: *** [CMakeFiles/Makefile2:892: googlemock/gtest/CMakeFiles/gtest_unittest.dir/all] Error 2
make[1]: Leaving directory &apos;/home/tkloczko/rpmbuild/BUILD/googletest-release-1.8.1/x86_64-redhat-linux-gnu&apos;
make: *** [Makefile:144: all] Error 2
make: Leaving directory &apos;/home/tkloczko/rpmbuild/BUILD/googletest-release-1.8.1/x86_64-redhat-linux-gnu&apos;
error: Bad exit status from /var/tmp/rpm-tmp.q9Gpv0 (%build)
</pre>

BTW: any chance than current base code will be released and tagged?

**Comment by kloczek** [comment]
With -Dgtest_build_tests=OFF and -Dgmock_build_tests=ON nuild fails on:
<pre>[ 16%] <font color="#4E9A06"><b>Linking CXX executable gmock-matchers_test</b></font>
cd /home/tkloczko/rpmbuild/BUILD/googletest-release-1.8.1/x86_64-redhat-linux-gnu/googlemock &amp;&amp; /usr/bin/cmake -E cmake_link_script CMakeFiles/gmock-matchers_test.dir/link.txt --verbose=1
/usr/bin/g++  -O2 -g -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -fstack-protector-strong -grecord-gcc-switches -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1 -m64 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection -flto  -Wl,-z,relro   -Wl,-z,now -specs=/usr/lib/rpm/redhat/redhat-hardened-ld -flto -fuse-linker-plugin -rdynamic CMakeFiles/gmock-matchers_test.dir/test/gmock-matchers_test.cc.o  -o gmock-matchers_test -Wl,-rpath,/home/tkloczko/rpmbuild/BUILD/googletest-release-1.8.1/x86_64-redhat-linux-gnu/googlemock:/home/tkloczko/rpmbuild/BUILD/googletest-release-1.8.1/x86_64-redhat-linux-gnu/googlemock/gtest libgmock_main.so.1.8.1 libgmock.so.1.8.1 gtest/libgtest.so.1.8.1 -pthread 
/home/tkloczko/rpmbuild/BUILD/googletest-release-1.8.1/googlemock/test/gmock-matchers_test.cc: In function &apos;__static_initialization_and_destruction_0.constprop&apos;:
/home/tkloczko/rpmbuild/BUILD/googletest-release-1.8.1/googlemock/test/gmock-matchers_test.cc:6805:1: note: variable tracking size limit exceeded with &apos;-fvar-tracking-assignments&apos;, retrying without
 6805 | }  // namespace testing
      | ^
/usr/bin/ld: libgmock.so.1.8.1: undefined reference to `testing::internal::ParseInt32(testing::Message const&amp;, char const*, int*)&apos;
collect2: error: ld returned 1 exit status
make[2]: *** [googlemock/CMakeFiles/gmock-matchers_test.dir/build.make:90: googlemock/gmock-matchers_test] Error 1
</pre>

**Comment by kloczek** [comment]
With -Dgtest_build_tests=OFF -Dgmock_build_tests=OFF and  -Dgtest_build_samples=ON build fails on:
<pre>[ 16%] <font color="#4E9A06"><b>Linking CXX executable gmock-matchers_test</b></font>
cd /home/tkloczko/rpmbuild/BUILD/googletest-release-1.8.1/x86_64-redhat-linux-gnu/googlemock &amp;&amp; /usr/bin/cmake -E cmake_link_script CMakeFiles/gmock-matchers_test.dir/link.txt --verbose=1
/usr/bin/g++  -O2 -g -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -fstack-protector-strong -grecord-gcc-switches -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1 -m64 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection -flto  -Wl,-z,relro   -Wl,-z,now -specs=/usr/lib/rpm/redhat/redhat-hardened-ld -flto -fuse-linker-plugin -rdynamic CMakeFiles/gmock-matchers_test.dir/test/gmock-matchers_test.cc.o  -o gmock-matchers_test -Wl,-rpath,/home/tkloczko/rpmbuild/BUILD/googletest-release-1.8.1/x86_64-redhat-linux-gnu/googlemock:/home/tkloczko/rpmbuild/BUILD/googletest-release-1.8.1/x86_64-redhat-linux-gnu/googlemock/gtest libgmock_main.so.1.8.1 libgmock.so.1.8.1 gtest/libgtest.so.1.8.1 -pthread 
/home/tkloczko/rpmbuild/BUILD/googletest-release-1.8.1/googlemock/test/gmock-matchers_test.cc: In function &apos;__static_initialization_and_destruction_0.constprop&apos;:
/home/tkloczko/rpmbuild/BUILD/googletest-release-1.8.1/googlemock/test/gmock-matchers_test.cc:6805:1: note: variable tracking size limit exceeded with &apos;-fvar-tracking-assignments&apos;, retrying without
 6805 | }  // namespace testing
      | ^
/usr/bin/ld: libgmock.so.1.8.1: undefined reference to `testing::internal::ParseInt32(testing::Message const&amp;, char const*, int*)&apos;
collect2: error: ld returned 1 exit status
make[2]: *** [googlemock/CMakeFiles/gmock-matchers_test.dir/build.make:90: googlemock/gmock-matchers_test] Error 1
</pre>

**Comment by EricWF** [comment]
The bug is in your compiler, no?

**Comment by kloczek** [comment]
I don't think that it is compiler bug.
As you see I'm using LTO and --as-needed.

**Comment by EricWF** [comment]
Oh, sorry. I misread the error. I thought we were talking about the '-fvar-tracking-assignments' diagnostic.

I'll look into this more.

**Comment by EricWF** [comment]
Can you provide your full CMake configuration command and output?
It would save me a much of time trying to reproduce.

Also, I don't see `--as-needed` in any of the linker invocations. You mentioned `-flto`, do you know if it links without it?

**Comment by kloczek** [comment]
That is setup which is working for me now.

<pre>Executing(%prep): /bin/sh -e /var/tmp/rpm-tmp.Hi9MH6
+ umask 022
+ cd /home/tkloczko/rpmbuild/BUILD
+ cd /home/tkloczko/rpmbuild/BUILD
+ rm -rf googletest-release-1.8.1
+ /usr/bin/gzip -dc /home/tkloczko/rpmbuild/SOURCES/release-1.8.1.tar.gz
+ /usr/bin/tar -xof -
+ STATUS=0
+ &apos;[&apos; 0 -ne 0 &apos;]&apos;
+ cd googletest-release-1.8.1
+ /usr/bin/chmod -Rf a+rX,u+w,g-w,o-w .
+ /usr/bin/cat /home/tkloczko/rpmbuild/SOURCES/gtest-null-pointer.patch
+ /usr/bin/patch -p1 -s --fuzz=0 --no-backup-if-mismatch
+ /usr/bin/cat /home/tkloczko/rpmbuild/SOURCES/gtest-PR1839-Fix-Python3-support.patch
+ /usr/bin/patch -p1 -s --fuzz=0 --no-backup-if-mismatch
+ /usr/bin/cat /home/tkloczko/rpmbuild/SOURCES/gtest-libversion.patch
+ /usr/bin/patch -p1 -s --fuzz=0 --no-backup-if-mismatch
+ /usr/bin/cat /home/tkloczko/rpmbuild/SOURCES/gtest-add-missing-pkgconfig-requires.patch
+ /usr/bin/patch -p1 -s --fuzz=0 --no-backup-if-mismatch
+ /usr/bin/cat /home/tkloczko/rpmbuild/SOURCES/gtest-python_path.patch
+ /usr/bin/patch -p1 -s --fuzz=0 --no-backup-if-mismatch
+ sed -e &apos;s/set(GOOGLETEST_VERSION .*)/set(GOOGLETEST_VERSION 1.8.1)/&apos; -i CMakeLists.txt
+ exit 0
Executing(%build): /bin/sh -e /var/tmp/rpm-tmp.hycW46
+ umask 022
+ cd /home/tkloczko/rpmbuild/BUILD
+ cd googletest-release-1.8.1
+ export &apos;CFLAGS=-O2 -g -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -fstack-protector-strong -grecord-gcc-switches -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1 -m64 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection -flto&apos;
+ CFLAGS=&apos;-O2 -g -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -fstack-protector-strong -grecord-gcc-switches -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1 -m64 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection -flto&apos;
+ export &apos;CXXFLAGS=-O2 -g -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -fstack-protector-strong -grecord-gcc-switches -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1 -m64 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection -flto -fno-rtti -fno-exceptions&apos;
+ CXXFLAGS=&apos;-O2 -g -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -fstack-protector-strong -grecord-gcc-switches -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1 -m64 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection -flto -fno-rtti -fno-exceptions&apos;
+ export &apos;FFLAGS=-O2 -g -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -fstack-protector-strong -grecord-gcc-switches -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1 -m64 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection -I/usr/lib64/gfortran/modules -flto&apos;
+ FFLAGS=&apos;-O2 -g -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -fstack-protector-strong -grecord-gcc-switches -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1 -m64 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection -I/usr/lib64/gfortran/modules -flto&apos;
+ export &apos;FCFLAGS=-O2 -g -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -fstack-protector-strong -grecord-gcc-switches -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1 -m64 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection -I/usr/lib64/gfortran/modules -flto&apos;
+ FCFLAGS=&apos;-O2 -g -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -fstack-protector-strong -grecord-gcc-switches -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1 -m64 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection -I/usr/lib64/gfortran/modules -flto&apos;
+ export &apos;LDFLAGS=-Wl,-z,relro -Wl,--as-needed  -Wl,-z,now -specs=/usr/lib/rpm/redhat/redhat-hardened-ld -flto -fuse-linker-plugin&apos;
+ LDFLAGS=&apos;-Wl,-z,relro -Wl,--as-needed  -Wl,-z,now -specs=/usr/lib/rpm/redhat/redhat-hardened-ld -flto -fuse-linker-plugin&apos;
+ export AR=gcc-ar RANLIB=gcc-ranlib NM=gcc-nm
+ AR=gcc-ar
+ RANLIB=gcc-ranlib
+ NM=gcc-nm
+ export CC=gcc CXX=g++
+ CC=gcc
+ CXX=g++
+ /usr/bin/cmake -DBUILD_SHARED_LIBS:BOOL=ON -DCMAKE_AR=/usr/bin/gcc-ar -DCMAKE_C_FLAGS_RELEASE:STRING=-DNDEBUG -DCMAKE_CXX_FLAGS_RELEASE:STRING=-DNDEBUG -DCMAKE_Fortran_FLAGS_RELEASE:STRING=-DNDEBUG -DCMAKE_INSTALL_PREFIX:PATH=/usr -DCMAKE_NM=/usr/bin/gcc-nm -DCMAKE_RANLIB=/usr/bin/gcc-ranlib -DCMAKE_VERBOSE_MAKEFILE:BOOL=ON -DINCLUDE_INSTALL_DIR:PATH=/usr/include -DLIB_INSTALL_DIR:PATH=/usr/lib64 -DLIB_SUFFIX=64 -DSHARE_INSTALL_PREFIX:PATH=/usr/share -DSYSCONF_INSTALL_DIR:PATH=/etc . -DBUILD_SHARED_LIBS=ON -Dgmock_build_tests=OFF -Dgtest_build_samples=OFF -Dgtest_build_tests=OFF -Dgtest_disable_pthreads=OFF -Dgtest_hide_internal_symbols=ON -DPYTHON_EXECUTABLE=/usr/bin/python3
-- The C compiler identification is GNU 9.1.1
-- The CXX compiler identification is GNU 9.1.1
-- Check for working C compiler: /usr/bin/gcc
-- Check for working C compiler: /usr/bin/gcc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /usr/bin/g++
-- Check for working CXX compiler: /usr/bin/g++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Found PythonInterp: /usr/bin/python3 (found version &quot;3.7.3&quot;) 
-- Looking for pthread.h
-- Looking for pthread.h - found
-- Looking for pthread_create
-- Looking for pthread_create - not found
-- Check if compiler accepts -pthread
-- Check if compiler accepts -pthread - yes
-- Found Threads: TRUE  
-- Configuring done
-- Generating done
CMake Warning:
  Manually-specified variables were not used by the project:

    CMAKE_CXX_FLAGS_RELEASE
    CMAKE_C_FLAGS_RELEASE
    CMAKE_Fortran_FLAGS_RELEASE
    INCLUDE_INSTALL_DIR
    LIB_INSTALL_DIR
    LIB_SUFFIX
    SHARE_INSTALL_PREFIX
    SYSCONF_INSTALL_DIR


</pre>

**Comment by EricWF** [comment]
Ooh. You're patching the source. That's important to know.

Before I go chasing my tail, have you tried building against trunk? 1.8.1 is quite old. It's very possible this bug has been fixed since then.

**Comment by kloczek** [comment]
Yest I have and without those patches build fails as well (I just tested that one more time to be sure).
If you are interested here are the patches which I'm using:
[gtest_patches.zip](https://github.com/google/googletest/files/3328524/gtest_patches.zip)

**Comment by EricWF** [comment]
The set of linker flags your passing is quite exotic.
For reference:

* '-Wl,-z,relro`
*`-Wl,--as-needed`
* `-Wl,-z,now`
* `-specs=/usr/lib/rpm/redhat/redhat-hardened-ld`
* `-flto`
* `-fuse-linker-plugin'
* `-rdynamic`

You're also building gtest and gmock as shared libraries, which doesn't play nice with LTO.
And as a further complication `-Dgtest_hide_internal_symbols=ON` is specified.

While looking into this issue I've found a dozen or so compiler or linker bugs that have the potential to be responsible. The most convincing of which was this one [1].

The above bug ties together LTO, the var tracking diagnostic, and the actual linker failure. It was last touched February 2019. I'm not sure if that fix made it into GCC 9.1.1

All of this being said, I'm certain this isn't a bug in Googletest. You're set up is essentially a stress test for bleeding edge compiler features -- which I know you're aware of because I found more than one of the compiler/linker bugs you filed related to LTO.

Is Googletest the only project you're running into issues building given your setup? Or are there other projects that fail to build?

In no particular order I would would try (A)  using static libraries instead of shared. (B) stop explicitly hiding the internal symbols, it could mess up LTO, and finally not using LTO at all?

I'm going to close this bug until there's sufficient evidence to suggest the bug we're witnessing is actually caused by incorrect code in Googletest.

[1]  https://patchwork.ozlabs.org/patch/711212/
[2] https://bugzilla.redhat.com/show_bug.cgi?id=1693831
